sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVFilters.scala

```diff
@@ -0,0 +1,212 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.csv
+
+import scala.util.Try
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.sources
+import org.apache.spark.sql.types.{BooleanType, StructType}
+
+/**
+ * An instance of the class compiles filters to predicates and allows to
+ * apply the predicates to an internal row with partially initialized values
+ * converted from parsed CSV fields.
+ *
+ * @param filters The filters pushed down to CSV datasource.
+ * @param dataSchema The full schema with all fields in CSV files.
+ * @param requiredSchema The schema with only fields requested by the upper layer.
+ * @param columnPruning true if CSV parser can read sub-set of columns otherwise false.
+ */
+class CSVFilters(
+    filters: Seq[sources.Filter],
+    dataSchema: StructType,
+    requiredSchema: StructType,
+    columnPruning: Boolean) {
+  require(checkFilters(), "All filters must be applicable to the data schema.")
+
+  /**
+   * The schema to read from the underlying CSV parser.
+   * It combines the required schema and the fields referenced by filters.
+   */
+  val readSchema: StructType = {
+    if (columnPruning) {
+      val refs = filters.flatMap(_.references).toSet
+      val readFields = dataSchema.filter { field =>
+        requiredSchema.contains(field) || refs.contains(field.name)
+      }
+      StructType(readFields)
+    } else {
+      dataSchema
+    }
+  }
+
+  /**
+   * Converted filters to predicates and grouped by maximum field index
+   * in the read schema. For example, if an filter refers to 2 attributes
+   * attrA with field index 5 and attrB with field index 10 in the read schema:
+   *   0 === $"attrA" or $"attrB" < 100
+   * the filter is compiled to a predicate, and placed to the `predicates`
+   * array at the position 10. In this way, if there is a row with initialized
+   * fields from the 0 to 10 index, the predicate can be applied to the row
+   * to check that the row should be skipped or not.
+   * Multiple predicates with the same maximum reference index are combined
+   * by the `And` expression.
+   */
+  private val predicates: Array[BasePredicate] = {
+    val len = readSchema.fields.length
+    val groupedPredicates = Array.fill[BasePredicate](len)(null)
+    if (SQLConf.get.csvFilterPushDown) {
+      val groupedExprs = Array.fill(len)(Seq.empty[Expression])
+      for (filter <- filters) {
+        val expr = CSVFilters.filterToExpression(filter, toRef)
+        val refs = filter.references
+        if (refs.isEmpty) {
+          // For example, AlwaysTrue and AlwaysFalse doesn't have any references
+          for (i <- 0 until len) {
+            groupedExprs(i) ++= expr
```

Comment:
for literals, shall we only put it in position 0? It only needs to be evaluated once.

Reply:
You are right since we combine all pushed filters via `And`. Also I think all filters with references (literals) could be put at the beginning of the group before reducing here https://github.com/apache/spark/pull/26973/files#diff-44a98c4a53980cb04e57f0489b257a37R95
So, we have pushed filters: 
`Seq(AlwaysFalse, StringContains(ref0, "abc"))`, and they are reduced to
`And(AlwaysFalse, StringContains(ref0, "abc"))`, the second filter `StringContains(ref0, "abc")` will not be evaluated at all.

Reply:
Even more, `AlwaysTrue` could be removed because it does not impact on the result. `AlwaysFalse` could be put at index 0, and other filters can be ignored.

Reply:
I have implemented the first approach in the commit: https://github.com/apache/spark/pull/26973/commits/c03ae069d738c6aa526cc1a1216d079bc8b5ec3e